### PR TITLE
Random Selfplay

### PR DESCRIPTION
**Pull request by vwxyzjn**
This PR prototypes fictitious selfplay. 

Known issue: still needs to change the following variables to exclude data from `agent2`

```
        b_obs = obs.reshape((-1,) + envs.observation_space.shape)
        b_logprobs = logprobs.reshape(-1)
        b_actions = actions.reshape((-1,) + action_space_shape)
        b_advantages = advantages.reshape(-1)
        b_returns = returns.reshape(-1)
        b_values = values.reshape(-1)
        b_invalid_action_masks = invalid_action_masks.reshape((-1,) + invalid_action_shape)
```

### CODE REVIEW

**Comment by vwxyzjn** [comment]
Running an experiment here: https://wandb.ai/costa-huang/gym-microrts/runs/x9y055w6

File: experiments/ppo_gridnet_league.py [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+                os.makedirs(f"models/{experiment_name}")
+            torch.save(agent.state_dict(), f"models/{experiment_name}/agent.pt")
+            torch.save(agent.state_dict(), f"models/{experiment_name}/{global_step}.pt")
+            if args.prod_mode:
+                wandb.save(f"models/{experiment_name}/agent.pt", base_path=f"models/{experiment_name}", policy="now")
+
+            # randomly load an opponent: fictitious self-play
+            list_of_agents = os.listdir(f"models/{experiment_name}")
+            list_of_agents.remove('agent.pt')
+            chosen_agent2pt = random.choice(list_of_agents)
```
Comment (kachayev) [inline experiments/ppo_gridnet_league.py:496]: Should we prune agents that have chance of winning < threshold? With the current implementation, it seems to that number of saved agent only grows over time, right? 🤔

Comment (kachayev) [reply experiments/ppo_gridnet_league.py:496]: Okay, I have much large question here. Will formulate it properly and will ask on Discord, so we can discuss 😀

Comment (vwxyzjn) [reply experiments/ppo_gridnet_league.py:496]: Haha yeah this first implementation is very crude. OpenAI Five does it by sampling opponents probabilistically according to their trueskill.

image

**Comment by vwxyzjn** [comment]
Per discussion with @kachayev, it turns out the implementation in this repo is definitely not fictitious selfplay, which also has a supervised learning component. Instead, this PR implements what I call "Random Selfplay" where the agent plays against a random past version of self.

**Comment by vwxyzjn** [comment]
Closed in favor of https://github.com/vwxyzjn/gym-microrts/pull/57

**Pull request closed without merging**